Feature/architecture refactor

AbdElrahman-Rafaat-Amer opened the pull request.

AbdElrahman-Rafaat-Amer committed:
create AuditCsvMapper.kt
create CsvDataSource.kt
create Exception.kt
change ProjectCsvMapper.kt
change TaskCsvMapper.kt
change UserCsvMapper.kt
change ProjectCsvMapper.kt

AbdElrahman-Rafaat-Amer committed:
create CreateProjectUseCase.kt
create CreateTaskUseCase.kt
create LogChangeUseCase.kt
change RegisterConsoleUI.kt
change RegisterPresenter.kt
change RegisterUseCase.kt

MahmoodTarek reviewed (approved):
amazing work Abdelrahman

DinaOthman21 commented:
Great Effort 👏

Mohamed-Elhanafy commented on src/main/kotlin/domain/model/Exception.kt:
```diff
@@ -0,0 +1,4 @@
+package org.example.domain.model
+
+class Exception {
```
we can rename this to EiffelFlowException

ahmednasserzaza commented on src/main/kotlin/data/storge/CsvMapper.kt:
```diff
@@ -0,0 +1,6 @@
+package org.example.data.storge
+
+interface CsvMapper<T> {
```
use this one instead to be reusable everywhere

`
interface Mapper<I, O> {
    fun map(input: I): O
}
`

ahmednasserzaza commented on src/main/kotlin/data/storge/DataSource.kt:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.UUID
+
+interface DataSource<T> {
```
I think using generics here is overengineering, just create all functions you need here like 
getTasks , saveTask , getProjects ...

ahmednasserzaza commented on src/main/kotlin/data/storge/mapper/TaskCsvMapper.kt:
```diff
@@ -0,0 +1,14 @@
+import org.example.data.storge.CsvMapper
+import org.example.domain.model.Task
+
+class TaskCsvMapper : CsvMapper<Task> {
```
After refactoring mapper as I mentioned before, here you can jsut override one function

ahmednasserzaza commented on src/main/kotlin/presentation/view/RegisterConsoleUI.kt:
```diff
@@ -0,0 +1,4 @@
+package org.example.presentation.view
+
+class RegisterConsoleUI {
```
Maybe RegisterCLI is better

ahmednasserzaza reviewed (approved):
Well done

AbdElrahman-Rafaat-Amer committed:
enhance the projects structure

remove CsvDataSource and make DataSource for each Actor like TaskDateSource.
create Mappers for all Actors
create Repositories for all Actors

Took 3 hours 56 minutes

AbdElrahman-Rafaat-Amer committed:
Merge remote-tracking branch 'origin/development' into feature/architecture-refactor

AbdElrahman-Rafaat-Amer committed:
fix the test cases imports

Took 7 minutes
